code/__HELPERS/files.dm

```diff
@@ -81,10 +81,38 @@
 /// Save file as an external file then md5 it.
 /// Used because md5ing files stored in the rsc sometimes gives incorrect md5 results.
 /proc/md5asfile(file)
+	var/filename = file2filepath(file)
+	var/istmp = FALSE
+	if (!filename || length(filename) < 1)
+		// its importaint this code can handle md5filepath sleeping instead of hard blocking, if it's converted to use rust_g.
+		filename = generate_tmp_filepath("md5asfile")
+		fcopy(file, filename)
+		istmp = TRUE
+	. = md5filepath(filename)
+	if (istmp)
+		fdel(filename)
+
+/// Generate a unique filepath inside the tmp folder.
+/// namespace - a optional string to prepend to the filename, useful for knowing what isn't cleaning up their things in the tmp folder
+/proc/generate_tmp_filepath(namespace = "tmpfile")
 	var/static/notch = 0
-	// its importaint this code can handle md5filepath sleeping instead of hard blocking, if it's converted to use rust_g.
-	var/filename = "tmp/md5asfile.[world.realtime].[world.timeofday].[world.time].[world.tick_usage].[notch]"
+	. = "tmp/[namespace].[world.realtime].[world.timeofday].[world.time].[world.tick_usage].[notch]"
```

Comment:
What's up with `notch`?

Reply:
if this is called multiple times in the same millisecond it will always lead to the same filename. notch serves the same function as the `sequence number` in twitter/discord snowflake ids.

I avoid the `2^24` limit by just making it wrap around.